File: src/main/scala/com/spark/example/DFTranspose.scala
package com.spark.example

import org.apache.spark.internal.Logging
import org.apache.spark.sql.functions.{col, collect_list, concat_ws}
import org.apache.spark.sql.{DataFrame, SparkSession}

object DFTranspose extends Logging {
  def main(args: Array[String]): Unit = {
    val spark: SparkSession = SparkSession.builder().appName("Spark DataFrame Transpose").master("local[*]").getOrCreate()
    val DataSeq = Seq(("Shirts", 10, 13, 34, 10), ("Trousers", 11, 2, 30, 20), ("Pants", 70, 43, 24, 60), ("Sweater", 101, 44, 54, 80))
    import spark.implicits._
    val productQtyDF = DataSeq.toDF("Products", "Small", "Medium", "Large", "ExLarge")

    log.info("Product Qty DataFrame (Input DataFrame)")
    productQtyDF.show(false)

    val productTypeDF = TransposeDF(productQtyDF, Seq("Small", "Medium", "Large", "ExLarge"), "Products")

    log.info("Product Type DataFrame (Transposed DataFrame)")
    productTypeDF.show(false)

  }


  def TransposeDF(df: DataFrame, columns: Seq[String], pivotCol: String): DataFrame = {
    val columnsValue = columns.map(x => "'" + x + "', " + x)
    val stackCols = columnsValue.mkString(",")
    val df_1 = df.selectExpr(pivotCol, "stack(" + columns.size + "," + stackCols + ")")
      .select(pivotCol, "col0", "col1")

    val final_df = df_1.groupBy(col("col0")).pivot(pivotCol).agg(concat_ws("", collect_list(col("col1"))))
      .withColumnRenamed("col0", pivotCol)
    final_df
  }

}

File: src/main/scala/com/spark/example/DFTranspose.py
from pyspark.sql.functions import *
from pyspark.sql import SparkSession



def TransposeDF(df, columns, pivotCol):
...     columnsValue = list(map(lambda x: str("'") + str(x) + str("',")  + str(x), columns))
...     stackCols = ','.join(x for x in columnsValue)
...     df_1 = df.selectExpr(pivotCol, "stack(" + str(len(columns)) + "," + stackCols + ")")\
...              .select(pivotCol, "col0", "col1")
...     final_df = df_1.groupBy(col("col0")).pivot(pivotCol).agg(concat_ws("", collect_list(col("col1"))))\
...                    .withColumnRenamed("col0", pivotCol)
...     return final_df

def main():
    DataList = [("Shirts", 10, 13, 34, 10), ("Trousers", 11, 2, 30, 20), ("Pants", 70, 43, 24, 60), ("Sweater", 101, 44, 54, 80)]
    productQtyDF  = spark.createDataFrame(DataList, ["Products", "Small", "Medium", "Large", "ExLarge"])
    productQtyDF.show()
    #+--------+-----+------+-----+-------+
    #|Products|Small|Medium|Large|ExLarge|
    #+--------+-----+------+-----+-------+
    #|  Shirts|   10|    13|   34|     10|
    #|Trousers|   11|     2|   30|     20|
    #|   Pants|   70|    43|   24|     60|
    #| Sweater|  101|    44|   54|     80|
    #+--------+-----+------+-----+-------+



    productTypeDF = TransposeDF(productQtyDF, ["Small", "Medium", "Large", "ExLarge"], "Products")
    productTypeDF.show(truncate= False)
    #+--------+-----+------+-------+--------+
    #|Products|Pants|Shirts|Sweater|Trousers|
    #+--------+-----+------+-------+--------+
    #|Medium  |43   |13    |44     |2       |
    #|Small   |70   |10    |101    |11      |
    #|ExLarge |60   |10    |80     |20      |
    #|Large   |24   |34    |54     |30      |
    #+--------+-----+------+-------+--------+


if __name__ == '__main__':
    main()
